architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+
+There are a few simple rules for this:
+
+- For operations on parent resources...
+  - A `DELETE` operation on a parent resource deletes all children.
+
+- For operations on child resources...
+  - A *`CRUL`* operation (anything but delete) on a child resource must return a `404` when the parent does not exist.
+  - A `DELETE` operation on a child resource never returns a `404`, because `DELETE` operations never return `404` in our API.
+
+### API design
+
+We should consider the set of concepts that need to be included in the resource provider and resource type APIs. For this design doc we'll consider the minimal set necessary to route to one of our existing resource providers. As we build more features for user-defined types we'll add more functionality. We don't want to design functionality in the API separately from its usage -- an incremental approach is better.
+
+The set of concepts needed right now looks like this:
+
+Resource providers:
+
+- (child resource): resource types
+- (child resource): locations
+
+Resource types:
+
+- (child resource): API versions
+- The default API version: the API version that should by used by a client that has no coupling to the API. This will mostly be internal UCP functionality like resource group deletion.
+
+API Versions:
+
+- ... (silence) ...
+- For now this will be empty. We'll add more functionality in the future as we build the rest of user-defined types.
+
+Locations:
+
+- An optional address/URL of the resource provider. This is optional because it won't be needed for declarative user-defined types.
+- The list of resource types supported in the location.
+  - The list of API versions supported for the resource type in the location
+
+Note: A resource type and API version may not be supported in all locations where the resource providers exists. This allows locations to be specialized based on functionality. This also allows for staged rollouts of new types.
+
+```typespec
+@doc("The resource provider namespace")
+@maxLength(63)
+@pattern("^([A-Za-z]([-A-Za-z0-9]*[A-Za-z0-9]))\\.([A-Za-z]([-A-Za-z0-9]*[A-Za-z0-9]))?$")
+scalar ResourceProviderNamespaceString extends string;
+
+@doc("The resource type name")
+@maxLength(63)
+@pattern("^([A-Za-z]([-A-Za-z0-9]*[A-Za-z0-9]))$")
+scalar ResourceTypeNameString extends string;
+
+@doc("The resource type name")
+@maxLength(63)
+@pattern("^(\\d{4}-\\d{2}-\\d{2}(-preview)?$")
+scalar ApiVersionNameString extends string;
+
+@doc("The location name")
+@maxLength(63)
+@pattern("^([A-Za-z]([-A-Za-z0-9]*[A-Za-z0-9]))$")
+scalar LocationNameString extends string;
+
+#suppress "@azure-tools/typespec-azure-resource-manager/arm-resource-path-segment-invalid-chars"
+model ResourceProviderResource is TrackedResource<ResourceProviderProperties> {
+  @key("resourceProviderName")
+  @doc("The resource provider name.")
+  @path
+  @segment("providers/System.Resources/resourceproviders")
```

Comment:
```suggestion
  @segment("providers/System.Resources/resourceProviders")
```

Reply:
The way we're handling OpenAPI unfortunately doesn't support this. I'm not a fan, but it has to be like this. 

Someone could work on that bug if they're motivated 😆